### DISCUSSION TITLE
Impove TypeScript Documentation

### DISCUSSION THREAD

**Discussion by MonstraG** [vercel/next.js#26811] [Ideas]
### Describe the feature you'd like to request

Based on [this](https://github.com/vercel/next.js/discussions/23445) (not mine).

Current [typescript tutorial for fetching data](https://nextjs.org/docs/basic-features/typescript#static-generation-and-server-side-rendering) is inadequate.

Right now I'm attempting to rewrite GetInitialProps typed like this:
```ts
type PageProps = {
    key: string;
}

const Page: NextPage<PageProps> = (props) => {...}

Page.getInitialProps = async (ctx: CustomContext): Promise<Props> => {...}

(class UserNextPageContext implements NextPageContext)
```
(And some other more complicated stuff)

And I don't have any idea how to write it.

### Describe the solution you'd like

Basically, provide working template in the documentation, where I can just replace Props with my own type and have it work or something.

### Describe alternatives you've considered

I've tried to follow [this part of the documentation](https://nextjs.org/docs/basic-features/data-fetching#typescript-use-getserversideprops)

but it didn't work:

![image](https://user-images.githubusercontent.com/17963791/117645632-3bbe1d00-b1a4-11eb-9a89-703a92ee0c86.png)

Only by adding NextPage<PageProps> (and removing suggested `Infer...`), it correctly resolved:
![image](https://user-images.githubusercontent.com/17963791/117645673-47a9df00-b1a4-11eb-911c-52c0bb5683ba.png)

Also, example doesn't fill out getServerSideProps return type, which in my case is `Promise<{ props: PageProps }>`, and `context` type, which is `GetServerSidePropsContext`.

And I'm not even talking about extending context in some way, just the default usage.
(The editor is Webstorm 2021.1, Typescript 4.2.4)

**Comment by RobbyUitbeijerse** [comment]
@MonstraG not sure if this helps, but note that you are able to type the return value of getServerSideProps/getStaticProps using the GetServerSidePropsResult/GetStaticPropsResult type utils

For example:

```
export type PageProps = PublicCountedVacancies & LayoutProps;

const Page: NextPage<PageProps> = ({ total, filters, vacancies }) => {
   return (...
}

export const getServerSideProps: GetServerSideProps = async (): Promise<
  GetServerSidePropsResult<PageProps>
> => {
   const data = await getItems()
   
   return {
       props: {
           ....
        }
     }
 }
```

hope this helps

**Comment by MonstraG** [comment]
> ... type the return value of getServerSideProps/getStaticProps using the GetServerSidePropsResult/GetStaticPropsResult type utils

That's what I'm talking about, that should be in documentation.

**Comment by ljosberinn** [comment]
> That's what I'm talking about, that should be in documentation.

This is a typescript issue, not nextjs. your screenshot shows you were not following the documentation 1:1 as evident here:

![image](https://user-images.githubusercontent.com/29307652/118143610-d5d3ce80-b40b-11eb-9892-4a15609d0b93.png)

The documentation mentions:

![image](https://user-images.githubusercontent.com/29307652/118143632-db311900-b40b-11eb-8bad-5828ae373ced.png)

So by adding the type for `GetServerSideProps` without passing it a type- which is solely typescript knowledge - youre forcing its return type [to whats defined within the `GetServerSidePropsType` ](https://github.com/vercel/next.js/blob/d5de1c53c1/packages/next/types/index.d.ts#L158), which is 
```ts
P extends { [key: string]: any } = { [key: string]: any }
```

Arguably, it would make more sense for next to not pass a default here, so you'd be forced to pass it a type.

If you remove `GetServerSideProps` there, it will infer properly.

**Comment by MonstraG** [comment]
> This is a typescript issue, not nextjs. your screenshot shows you were not following the documentation 1:1 as evident here:
> ...
> If you remove GetServerSideProps there, it will infer properly.

Documentation [pretty clearly states](https://nextjs.org/docs/basic-features/data-fetching#typescript-use-getserversideprops)

to use `GetServerSideProps`, and `InferGetServerSidePropsType<typeof getServerSideProps>`
